keps/sig-node/3008-cri-class-based-resources/README.md

```diff
@@ -0,0 +1,1606 @@
+possible to configure pod sandbox-level cgroup parameters). However, having
+pod-level QoS-class resources makes it possible to implement support for
+sandbox-level blockio parameters. Other usage for pod sandbox-level QoS-class
+resources would be communicating the Kubernetes Pod QoS class from
+kubelet to the container runtime.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- Make it possible to request QoS-class resources
+  - Support RDT class assignment of containers. This is already supported by
+    the containerd and CRI-O runtime and part of the OCI runtime-spec
+  - Support blockio class assignment of containers.
+  - Support Pod-level (sandbox-level) QoS-class resources
+- Make the API to support updating QoS-class resource assignment of running containers
+- Make the extensions flexible, enabling simple addition of other QoS-class
+  resource types in the future.
+- Make QoS-class resources opqaue (as possible) to the CRI client
+- Discovery of the available QoS-class resources
+- API changes to support updating Pod-level (sandbox-level) QoS-class resource
+  assignment of running pods ([future work](#future-work))
+- Resource status/capacity ([future work](#future-work))
+- Access control ([future work](#future-work))
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+- Interface or mechanism for configuring the QoS-class resources (responsibility of
+  the container runtime).
+- Enumerating possible (QoS-class) resource types or their detailed behavior
+
+## Implementation phases
+
+This proposal splits the full implementation of QoS-class resources into
+multiple phases, building functionality gradually, step-by-step. The goal is to
+make the discussions more focused and easier. We may also learn on the way,
+insights from earlier phases affecting design choises made in the later phases,
+hopefully resulting in a better overall end result. However, we also outline
+all the future steps to not lose the overall big picture.
+
+The figure below illustrates the design of the full implementation (less quota)
+and the division of implementation phases. The first implementation phase
+basically covers the communication between kubelet and the container runtime
+(i.e. CRI API). All changes to the Kubernetes API and its control plane
+components are left to future work. This KEP (the [Proposal](#proposal) and
+[Design Details](#design-details)) in its current form implements this first
+phase – the KEP will evolve and be supplemented with future phases getting
+implemented.
+
+![design](./design.svg)
+
+In the current design QoS-class resources are designed to be opaque to the CRI
+client in the sense that the container runtime takes care of configuration and
+control of the resources and the classes within.
+
+### Phase 1
```

Comment:
If the end-to-end design can be implemented via annotations alone, can we flesh this concept out further outside of Kubernetes?  Once we see successful application of the concept, we can then explore the API changes (phase 2+) with more known/known use cases that would benefit?  It feels like phase 2+ is where the KEP intersects more with Kubernetes itself.

Reply:
I don't know what to say. We have explored this outside Kubernetes and there are customers using this experimental runtime-specific mechanism (for RDT and blockio).

E.g. regarding RDT: OCI runtime spec supports it, runtimes support it -> I think the question should be how Kubernetes should support it. It could be e.g. decicate field (say `RdtClass`) in CRI and PodSpec et at. However, we thought this is too short-sighted as we could see the need for other similar "QoS resources". Thus, in this proposal we suggest a generic mechanism to support all these. But that can be questioned, of course